packages/firestore/src/lite-api/query.ts

```diff
@@ -193,6 +201,92 @@ export function where(
   return new QueryFilterConstraint(field, op, value);
 }
 
+class QueryCompositeFilterConstraint extends QueryConstraint {
+  constructor(
+    readonly type: 'or' | 'and',
+    private readonly _queryConstraints: QueryFilterConstraint[]
+  ) {
+    super();
+  }
+
+  _parse<T>(query: Query<T>): Filter {
+    const parsedFilters = this._queryConstraints
+      .map(queryConstraint => {
+        return queryConstraint._parse(query);
+      })
+      .filter(parsedFilter => parsedFilter.getFilters().length > 0);
+
+    if (parsedFilters.length === 1) {
+      return parsedFilters[0];
+    }
+
+    return CompositeFilter.create(parsedFilters, this.getOperator());
+  }
+
+  _apply<T>(query: Query<T>): Query<T> {
+    const parsedFilter = this._parse(query);
+    if (parsedFilter.getFilters().length === 0) {
+      // Return the existing query if not adding any more filters (e.g. an empty composite filter).
+      return query;
+    }
+    validateNewFilter(query._query, parsedFilter);
+
+    return new Query(
+      query.firestore,
+      query.converter,
+      queryWithAddedFilter(query._query, parsedFilter)
+    );
+  }
+
+  getQueryConstraints(): QueryConstraint[] {
+    return this._queryConstraints;
+  }
+
+  getOperator(): CompositeOperator {
+    return this.type === 'and' ? CompositeOperator.AND : CompositeOperator.OR;
+  }
+}
+
+/**
+ * Creates a {@link QueryConstraint} that performs a logical OR
+ * of all the provided `queryConstraints`
+ *
+ * @param queryConstraints - Optional. The {@link queryConstraints} for OR operation. These must be
+ * created with calls to {@link where}, {@link or}, or {@link and}.
+ * @returns The created {@link QueryConstraint}.
+ */
+export function or(...queryConstraints: QueryConstraint[]): QueryConstraint {
```

Comment:
is there a way to _**not**_ expose `and()` and `or()` publicly yet (and still be able to invoke them in the tests) ?

Android makes that possible using `@RestrictTo(RestrictTo.Scope.LIBRARY)`. But not sure if JS has a way to do this.

Reply:
My understanding is by keeping it out of [api.ts](https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/api.ts), it's still internal, but can be used for testing. Please correct me if you see a gap in my understanding.

Reply:
Yep; just looked into it a bit more. (as a side, we also have ["lite" APIs](https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/lite/index.ts) for the ["Lite Web SDK"](https://firebase.google.com/docs/firestore/solutions/firestore-lite)).

Reply:
@MarkDuckworth , IIUC based on @hsubox76 's comment, this comment still stands